src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      for (const key in action.extraData) {
+        result[key] = action.extraData[key];
+      }
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete result[key];
+      }
+    }
+
+    if (action.type === 'clear') {
+      for (const key in result) {
+        delete result[key];
+      }
+    }
+    stateHistory.push({ ...result });
```

Comment:
Cloning the `result` object before pushing it to `stateHistory` is a good practice to ensure that each state in the history is independent of future modifications.